pkg/infra/baseagent.go

```diff
@@ -93,8 +95,39 @@ func (a *Base) setHP(newHP int) {
 	a.hp = newHP
 }
 
+// Modeled as a first order system step answer (see documentation for more information)
 func (a *Base) updateHP(foodTaken float64) {
-	a.hp = int(math.Min(100, float64(a.hp)+foodTaken))
+	tau := 0.0
+	base := 0
+	width := 0
+
+	switch {
+	case a.hp >= a.tower.healthInfo.StrongLevel:
+		tau = a.tower.healthInfo.TauStrong
+		base = a.tower.healthInfo.StrongLevel
+		width = a.tower.healthInfo.WidthStrong
+
+	case a.hp >= a.tower.healthInfo.HealthyLevel:
+		tau = a.tower.healthInfo.TauHealthy
+		base = a.tower.healthInfo.HealthyLevel
+		width = a.tower.healthInfo.WidthHealthy
+
+	case a.hp >= a.tower.healthInfo.WeakLevel:
+		tau = a.tower.healthInfo.TauWeak
+		base = a.tower.healthInfo.WeakLevel
+		width = a.tower.healthInfo.WidthWeak
+
+	// Critical Level - TODO: discuss its implementation
+	case a.hp >= 0:
+		if foodTaken >= a.tower.healthInfo.FoodReqCToW {
+			a.hp = int(a.tower.healthInfo.WeakLevel) - 1
+		} else {
+			a.hp = 1
+		}
+		return
+	}
+
+	a.hp = base + width - int(math.Pow(math.E, -foodTaken/tau)*float64(width-a.hp+base))
```

Comment:
If you are midway through a level and then eat the value of tau, you will end the day on less hp than you started. I don't think this is the correct way to implement it. The shifting of the curve as described in the docs means that moving up to a new level is too difficult in my opinion. I give two example below to show why I think this is an issue.

Example 1:
You start at 45 hp
You eat 15 hp (tau in healthy), taking you to 51 hp by the curve.
In hpDecay, you lose 19 hp (0.63*30) taking you to 32 hp at the end of the day.

Example 2:
You start at 45 hp
You eat 20 hp (the value of tau in strong), taking you to 52 hp by the curve
In hp decay, you lose 19 hp taking you to 33 hp at the end of the day.
This means that maintaining hp in the upper section of the healthy range is more difficult than in the strong stage, which I don't think should be the case. 

I think this needs further discussion unless others disagree with this being an issue or I have completely misinterpreted the code, if this is the case then please correct me. Otherwise, I think this needs a rework.

Reply:
This is indeed a very good point, thank you for the examples; they help to understand your thoughts. 
The second example is for me the most relevant. I agree that maintaining your hp in the healthy level (independently to where you are exactly in this level) should not be more difficult than doing so in the strong level. I see two possible ways to fix this:
1) We shift the curves vertically instead of horizontally. In this case the diminishing return is only dependent on the amount of food (and your hp level as it parameterizes the curve), and _not_ on your exact hp value. 
2) Try to find a parametrisation of the hp levels that does not lead to such critical scenarios.

I think 1) makes quite a lot of sense to me. The original idea of shifting the curve horizontally was intended to capture the diminishing return as a function of your hp, but it makes actually more sense to have diminishing returns depending on the amount of food you eat. 1) would capture this.

Reply:
If we decide to implement 1), there is one additional thing we would have to discuss. Right now, the curves saturate at the upper border of each level. this was done so to prevent agents to jump from, for example, weak to strong in one day. Indeed, we want to model the fact that it needs time to "recover" and to feel better. 
If we shift the curves vertically, we would have curves that saturate higher than the level upper boundary. From this, two questions come into my mind:
a) what is the threshold to wake up in the upper level the following day? Is it still as soon as you reach the border of the level ?
b) how do we compute hpDecay if we allow curve to cross the level borders? Do we have something like `max{upperLevel, curentHP-0.63*levelWidth}`, where `levelWidth` is defined as the width of the level the agent started in?
c) What should we do with the amount of food taken by the agent after the curve cross the upper border? Does this food still contribute to increase the agent's hp or do we say something like "you reached the next hp level, you will wake up in this level but your extra food brings you nothing?

Reply:
I think shifting the curves vertically is the right way to do it in this implementation. Changing the max limit of hp you can get to from base+width to a.hp+width would set the upper limit to a more realistic threshold. Then the exponential will be multiplied by the width only. 

```suggestion
	a.hp = a.hp + width - int(math.Pow(math.E, -foodTaken/tau)*float64(width))
```

Maybe this would change the curve to a vertical shift. Not sure what changes need to happen in hpDecay for this to be acceptable though.

I think the limit for hp at the end of the day should not be set by the level thresholds, but by the starting hp plus some maximum gain (width)

Reply:
I just came back home, so I'll have a look at this in details tonight/tomorrow. Your suggestion seems good. We just need to clarify what what when an agent change his level.

Reply:
@Rhys-J I commit your suggestion. It makes sense to me. For updateHP, I also added a safeguard so that the HP can not grow bigger than maxHP.

Reply:
Concerning hpDecay, I left it as it is. It is almost good, but we still have a small problem. Right now, the "cost of living" is always `0.63*currentWidth` (with safeguard in the code to prevent falling too low-normaly not needed with good choice the levelWidth), except if `a.hp>=0.95*currentWidth`. In this last case, the cost of living is 0 and you go up one level. This seems a bit strange with the current vertical shift implementation, as _you could wake up at_ `a.hp=currentLevel+0.95*currentWidth`, _eat nothing, and wake up in the upper level the following day_.

We could avoid this by saying that the cost of living is _really always_ `0.63*currentWidth`, and don't treat this scenario where `a.hp>=0.95*currentWidth` separately. It seems appealing, but this would lead to the fact that it is would take several days for an agent to "level up" starting from the lower border of a level. Let me give you an example:

> You start at the lower border of healthy. **Day 1** you eat and get +0.83*widthHealthy hp. **Day 2** you wake up with `healthyLevel+0.83*healthyWidth-0.63*healthy Width=healthyLevel+0.20*width Healthy`. You eat and get +0.83*widthHealthy hp. **Day 3** You wake up at  `healthyLevel+0.2*healthyWidth+0.83*healthyWidth-0.63*strong Width<=healthyLevel+0.40*width Healthy`(as you substract **strong**Width) Etc etc. Also if you eat 100% of the currentWidth, it takes you a bit of time to level up. >

Therefore, I would probably keep it like this for now, but I am not completely happy with it

Reply:
**Paramatrization of tau**: I also changed the values of tau so that it makes more sense. When running simulations, I saw that with our former parametrisation, eating the same quantity of food in Strong level could actually bring you more hp than in the weak level. This is because the exponential decay term it not big enough and the Width factor in front of the parenthese is too large. 
In order to choose better values for tau, I kept the initial value of the level and plot the curves for different values of tau. I chose values that allow a large hp increase in the lower levels for a fixed amount of food.